diff -urN --no-dereference libinput-clean/include/linux/input.h libinput-workdir/include/linux/input.h
--- libinput-clean/include/linux/input.h
+++ libinput-workdir/include/linux/input.h
@@ -2,4 +2,6 @@
 #include "linux/input.h"
 #elif __FreeBSD__
 #include "freebsd/input.h"
+#elif __menix__
+#include_next <linux/input.h>
 #endif
diff -urN --no-dereference libinput-clean/meson.build libinput-workdir/meson.build
--- libinput-clean/meson.build
+++ libinput-workdir/meson.build
@@ -115,9 +115,9 @@
 	config_h.set('HAVE_VERSIONSORT', '1')
 endif
 
-if cc.get_define('SYS_pidfd_open', prefix: '#include <sys/syscall.h>') != ''
-	config_h.set('HAVE_PIDFD_OPEN', '1')
-endif
+# if cc.get_define('SYS_pidfd_open', prefix: '#include <sys/syscall.h>') != ''
+# 	config_h.set('HAVE_PIDFD_OPEN', '1')
+# endif
 
 if cc.has_function('sigabbrev_np', prefix: '#define _GNU_SOURCE 1\n#include <string.h>')
 	config_h.set('HAVE_SIGABBREV_NP', '1')
diff -urN --no-dereference libinput-clean/tools/libinput-debug-tablet.c libinput-workdir/tools/libinput-debug-tablet.c
--- libinput-clean/tools/libinput-debug-tablet.c
+++ libinput-workdir/tools/libinput-debug-tablet.c
@@ -33,6 +33,7 @@
 #include <string.h>
 #include <signal.h>
 #include <sys/ioctl.h>
+#include <termios.h>
 #include <unistd.h>
 #include <libinput.h>
 #include <libevdev/libevdev.h>
