diff --git libevdev-clean/include/linux/input.h libevdev-workdir/include/linux/input.h
index 03c512e..eac1212 100644
--- libevdev-clean/include/linux/input.h
+++ libevdev-workdir/include/linux/input.h
@@ -1,5 +1,5 @@
-#ifdef __linux__
+#if defined __linux__
 #include "linux/input.h"
-#elif __FreeBSD__
+#elif __FreeBSD__ || defined __menix__
 #include "freebsd/input.h"
 #endif
diff --git libevdev-clean/include/linux/uinput.h libevdev-workdir/include/linux/uinput.h
index 1ef4e3b..d6389b8 100644
--- libevdev-clean/include/linux/uinput.h
+++ libevdev-workdir/include/linux/uinput.h
@@ -1,5 +1,5 @@
-#ifdef __linux__
+#if defined __linux__
 #include "linux/uinput.h"
-#elif __FreeBSD__
+#elif __FreeBSD__ || defined __menix__
 #include "freebsd/uinput.h"
 #endif
