Index: busybox/include/platform.h
===================================================================
--- busybox.orig/include/platform.h
+++ busybox/include/platform.h
@@ -462,6 +462,11 @@ typedef unsigned smalluint;
 # undef HAVE_NET_ETHERNET_H
 #endif
 
+#if defined(__menix__)
+# undef HAVE_SYS_STATFS_H
+# undef HAVE_MNTENT_H
+#endif
+
 #if defined(__CYGWIN__)
 # undef HAVE_CLEARENV
 # undef HAVE_FDPRINTF
Index: busybox/libbb/appletlib.c
===================================================================
--- busybox.orig/libbb/appletlib.c
+++ busybox/libbb/appletlib.c
@@ -33,7 +33,7 @@ static inline int *get_perrno(void) { re
 #include "busybox.h"
 
 #if !(defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) \
-    || defined(__APPLE__) \
+    || defined(__APPLE__) || defined(__menix__) \
     )
 # include <malloc.h> /* for mallopt */
 #endif
Index: busybox/libbb/Kbuild.src
===================================================================
--- busybox.orig/libbb/Kbuild.src
+++ busybox/libbb/Kbuild.src
@@ -10,7 +10,7 @@ lib-y:=
 
 INSERT
 
-lib-y += alloc_affinity.o
+# lib-y += alloc_affinity.o
 lib-y += appletlib.o
 lib-y += ask_confirmation.o
 lib-y += bb_askpass.o
