diff -urN --no-dereference gcc-clean/fixincludes/mkfixinc.sh gcc-workdir/fixincludes/mkfixinc.sh
--- gcc-clean/fixincludes/mkfixinc.sh
+++ gcc-workdir/fixincludes/mkfixinc.sh
@@ -11,6 +11,7 @@
 
 # Check for special fix rules for particular targets
 case $machine in
+    *-mlibc* | \
     i?86-*-cygwin* | \
     i?86-*-mingw32* | \
     x86_64-*-mingw32* | \
diff -urN --no-dereference gcc-clean/gcc/config/aarch64/aarch64-menix.h gcc-workdir/gcc/config/aarch64/aarch64-menix.h
--- gcc-clean/gcc/config/aarch64/aarch64-menix.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-workdir/gcc/config/aarch64/aarch64-menix.h
@@ -0,0 +1,49 @@
+#undef GCC_AARCH64_MENIX
+#define GCC_AARCH64_MENIX 1
+
+#undef GNU_USER_DYNAMIC_LINKER
+#define GNU_USER_DYNAMIC_LINKER "/lib/ld.so"
+
+#define MENIX_TARGET_LINK_SPEC  "%{h*}                       \
+  %{static:-Bstatic}                                         \
+  %{shared:-shared}                                          \
+  %{symbolic:-Bsymbolic}                                     \
+  %{!static:%{!static-pie:                                   \
+    %{rdynamic:-export-dynamic}                              \
+    %{!shared:-dynamic-linker " GNU_USER_DYNAMIC_LINKER "}}} \
+  %{static-pie:-Bstatic -pie --no-dynamic-linker -z text}    \
+  -X                                                         \
+  -EL"
+
+#if TARGET_FIX_ERR_A53_835769_DEFAULT
+#define CA53_ERR_835769_SPEC \
+  " %{!mno-fix-cortex-a53-835769:--fix-cortex-a53-835769}"
+#else
+#define CA53_ERR_835769_SPEC \
+  " %{mfix-cortex-a53-835769:--fix-cortex-a53-835769}"
+#endif
+
+#if TARGET_FIX_ERR_A53_843419_DEFAULT
+#define CA53_ERR_843419_SPEC \
+  " %{!mno-fix-cortex-a53-843419:--fix-cortex-a53-843419}"
+#else
+#define CA53_ERR_843419_SPEC \
+  " %{mfix-cortex-a53-843419:--fix-cortex-a53-843419}"
+#endif
+
+#define LINK_SPEC MENIX_TARGET_LINK_SPEC \
+                  CA53_ERR_835769_SPEC \
+                  CA53_ERR_843419_SPEC
+
+#define GNU_USER_TARGET_MATHFILE_SPEC \
+  "%{Ofast|ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC   \
+  GNU_USER_TARGET_MATHFILE_SPEC " " \
+  GNU_USER_TARGET_ENDFILE_SPEC
+
+#define TARGET_OS_CPP_BUILTINS()       \
+  do {                                 \
+    GNU_USER_TARGET_OS_CPP_BUILTINS(); \
+  } while (0)
diff -urN --no-dereference gcc-clean/gcc/config/aarch64/t-aarch64-menix gcc-workdir/gcc/config/aarch64/t-aarch64-menix
--- gcc-clean/gcc/config/aarch64/t-aarch64-menix	1970-01-01 01:00:00.000000000 +0100
+++ gcc-workdir/gcc/config/aarch64/t-aarch64-menix
@@ -0,0 +1,3 @@
+
+LIB1ASMSRC   = aarch64/lib1funcs.asm
+LIB1ASMFUNCS = _aarch64_sync_cache_range
diff -urN --no-dereference gcc-clean/gcc/config/i386/menix.h gcc-workdir/gcc/config/i386/menix.h
--- gcc-clean/gcc/config/i386/menix.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-workdir/gcc/config/i386/menix.h
@@ -0,0 +1,10 @@
+#undef GCC_I386_MENIX
+#define GCC_I386_MENIX 1
+
+#define GNU_USER_LINK_EMULATION32 "elf_i386"
+#define GNU_USER_LINK_EMULATION64 "elf_x86_64"
+#define GNU_USER_LINK_EMULATIONX32 "elf32_x86_64"
+
+#define GNU_USER_DYNAMIC_LINKER64 "/lib/ld.so"
+#define GNU_USER_DYNAMIC_LINKER32 "/dev/null"
+#define GNU_USER_DYNAMIC_LINKERX32 "/dev/null"
diff -urN --no-dereference gcc-clean/gcc/config/loongarch/menix.h gcc-workdir/gcc/config/loongarch/menix.h
--- gcc-clean/gcc/config/loongarch/menix.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-workdir/gcc/config/loongarch/menix.h
@@ -0,0 +1,6 @@
+#undef GCC_LOONGARCH_MENIX
+#define GCC_LOONGARCH_MENIX 1
+
+#undef GNU_USER_DYNAMIC_LINKER
+#define GNU_USER_DYNAMIC_LINKER "/lib/ld.so"
+#define GNU_USER_LINK_EMULATION "elf" ABI_GRLEN_SPEC "loongarch"
diff -urN --no-dereference gcc-clean/gcc/config/menix.h gcc-workdir/gcc/config/menix.h
--- gcc-clean/gcc/config/menix.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-workdir/gcc/config/menix.h
@@ -0,0 +1,13 @@
+#undef TARGET_MENIX
+#define TARGET_MENIX 1
+
+#undef GNU_USER_TARGET_OS_CPP_BUILTINS
+#define GNU_USER_TARGET_OS_CPP_BUILTINS() \
+  do {                                    \
+    builtin_define ("__menix__");         \
+    builtin_define ("__mlibc__");         \
+    builtin_define ("__unix__");          \
+    builtin_assert ("system=menix");      \
+    builtin_assert ("system=unix");       \
+    builtin_assert ("system=posix");      \
+  } while (0);
diff -urN --no-dereference gcc-clean/gcc/config/riscv/menix.h gcc-workdir/gcc/config/riscv/menix.h
--- gcc-clean/gcc/config/riscv/menix.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-workdir/gcc/config/riscv/menix.h
@@ -0,0 +1,33 @@
+#undef GCC_RISCV_MENIX
+#define GCC_RISCV_MENIX 1
+
+#define ICACHE_FLUSH_FUNC "__riscv_flush_icache"
+
+#undef GNU_USER_DYNAMIC_LINKER
+#define GNU_USER_DYNAMIC_LINKER "/lib/ld.so"
+
+#define LD_EMUL_SUFFIX \
+  "%{mabi=lp64d:}" \
+  "%{mabi=lp64f:_lp64f}" \
+  "%{mabi=lp64:_lp64}" \
+  "%{mabi=ilp32d:}" \
+  "%{mabi=ilp32f:_ilp32f}" \
+  "%{mabi=ilp32:_ilp32}"
+
+#define LINK_SPEC "\
+-melf" XLEN_SPEC DEFAULT_ENDIAN_SPEC "riscv" LD_EMUL_SUFFIX " \
+%{mno-relax:--no-relax} \
+%{mbig-endian:-EB} \
+%{mlittle-endian:-EL} \
+%{shared} \
+  %{!shared: \
+    %{!static: \
+      %{!static-pie: \
+    %{rdynamic:-export-dynamic} \
+    -dynamic-linker " GNU_USER_DYNAMIC_LINKER "}} \
+    %{static:-static} %{static-pie:-static -pie --no-dynamic-linker -z text}}"
+
+#define TARGET_OS_CPP_BUILTINS()       \
+  do {                                 \
+    GNU_USER_TARGET_OS_CPP_BUILTINS(); \
+  } while (0)
diff -urN --no-dereference gcc-clean/gcc/config.build gcc-workdir/gcc/config.build
--- gcc-clean/gcc/config.build
+++ gcc-workdir/gcc/config.build
@@ -45,7 +45,7 @@
 build_xm_file=
 build_xm_defines=
 build_exeext=
-build_install_headers_dir=install-headers-tar
+build_install_headers_dir=install-headers-cp
 build_file_translate=
 
 # System-specific settings.
diff -urN --no-dereference gcc-clean/gcc/config.gcc gcc-workdir/gcc/config.gcc
--- gcc-clean/gcc/config.gcc
+++ gcc-workdir/gcc/config.gcc
@@ -187,7 +187,7 @@
 #			the --with-sysroot configure option or the
 #			--sysroot command line option is used this
 #			will be relative to the sysroot.
-# target_type_format_char 
+# target_type_format_char
 # 			The default character to be used for formatting
 #			the attribute in a
 #			.type symbol_name, ${t_t_f_c}<property>
@@ -884,6 +884,15 @@
   rust_target_objs="${rust_target_objs} fuchsia-rust.o"
   target_has_targetrustm=yes
   ;;
+*-*-*-mlibc)
+  extra_options="$extra_options gnu-user.opt"
+  gas=yes
+  gnu_ld=yes
+  default_use_cxa_atexit=yes
+  use_gcc_stdint=wrap
+  tmake_file="${tmake_file} t-slibgcc"
+  thread_file='posix'
+  ;;
 *-*-linux* | frv-*-*linux* | *-*-kfreebsd*-gnu | *-*-gnu* | *-*-kopensolaris*-gnu | *-*-uclinuxfdpiceabi)
   extra_options="$extra_options gnu-user.opt"
   gas=yes
@@ -1239,6 +1248,11 @@
 	tmake_file="${tmake_file} aarch64/t-aarch64 aarch64/t-aarch64-netbsd"
 	extra_options="${extra_options} netbsd.opt netbsd-elf.opt"
 	;;
+aarch64*-*-menix*)
+	tm_file="${tm_file} elfos.h gnu-user.h menix.h glibc-stdint.h"
+	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-menix.h"
+	tmake_file="${tmake_file} aarch64/t-aarch64 aarch64/t-aarch64-menix"
+	;;
 aarch64*-*-linux*)
 	tm_file="${tm_file} elfos.h gnu-user.h linux.h glibc-stdint.h"
 	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-errata.h aarch64/aarch64-linux.h"
@@ -2284,6 +2298,9 @@
 			;;
 	esac
 	;;
+x86_64-*-menix*)
+	tm_file="${tm_file} i386/unix.h i386/att.h elfos.h gnu-user.h menix.h glibc-stdint.h i386/x86-64.h i386/gnu-user-common.h i386/gnu-user64.h i386/menix.h"
+	;;
 x86_64-*-fuchsia*)
 	tmake_file="${tmake_file} i386/t-x86_64-elf"
 	tm_file="${tm_file} i386/unix.h i386/att.h elfos.h newlib-stdint.h i386/i386elf.h i386/x86-64.h fuchsia.h"
@@ -2483,6 +2500,13 @@
 	cxx_target_objs="${cxx_target_objs} microblaze-c.o"
 	tmake_file="${tmake_file} microblaze/t-microblaze"
         ;;
+riscv*-*-menix*)
+	tm_file="${tm_file} elfos.h gnu-user.h menix.h glibc-stdint.h riscv/menix.h"
+	tmake_file="${tmake_file} riscv/t-riscv"
+	gnu_ld=yes
+	gas=yes
+	gcc_cv_initfini_array=yes
+	;;
 riscv*-*-linux*)
 	tm_file="elfos.h gnu-user.h linux.h glibc-stdint.h ${tm_file} riscv/linux.h"
 	case "x${enable_multilib}" in
@@ -2564,6 +2588,15 @@
 	gcc_cv_initfini_array=yes
 	;;
 
+loongarch*-*-menix*)
+	tm_file="${tm_file} elfos.h gnu-user.h menix.h glibc-stdint.h"
+	tm_file="${tm_file} loongarch/gnu-user.h loongarch/menix.h loongarch/loongarch-driver.h"
+	tmake_file="${tmake_file} loongarch/t-multilib"
+	gnu_ld=yes
+	gas=yes
+	gcc_cv_initfini_array=yes
+	;;
+
 loongarch*-*-elf*)
 	tm_file="elfos.h newlib-stdint.h ${tm_file}"
 	tm_file="${tm_file} loongarch/elf.h loongarch/linux.h loongarch/loongarch-driver.h"
@@ -5062,6 +5095,7 @@
 		esac
 
 		case ${target} in
+		  *-menix-mlibc*) triplet_os="menix-mlibc";;
 		  *-linux-gnu*) triplet_os="linux-gnu";;
 		  *-linux-musl*) triplet_os="linux-musl";;
 		  *-elf*) triplet_os="elf";;
diff -urN --no-dereference gcc-clean/gcc/config.host gcc-workdir/gcc/config.host
--- gcc-clean/gcc/config.host
+++ gcc-workdir/gcc/config.host
@@ -61,7 +61,7 @@
 #			different from the default "liblto_plugin.so".
 
 # When setting any of these variables, check to see if a corresponding
-# variable is present in config.build; if so, you will likely want to 
+# variable is present in config.build; if so, you will likely want to
 # set it in both places.
 
 # Default settings.
@@ -100,7 +100,7 @@
 
 case ${host} in
   aarch64*-*-freebsd* | aarch64*-*-linux* | aarch64*-*-fuchsia* |\
-  aarch64*-*-darwin*)
+  aarch64*-*-darwin* | aarch64*-*-menix*)
     case ${target} in
       aarch64*-*-*)
 	host_extra_gcc_objs="driver-aarch64.o"
diff -urN --no-dereference gcc-clean/libatomic/configure.tgt gcc-workdir/libatomic/configure.tgt
--- gcc-clean/libatomic/configure.tgt
+++ gcc-workdir/libatomic/configure.tgt
@@ -151,7 +151,7 @@
   *-*-linux* | *-*-gnu* | *-*-k*bsd*-gnu \
   | *-*-netbsd* | *-*-freebsd* | *-*-openbsd* | *-*-dragonfly* \
   | *-*-solaris2* | *-*-sysv4* | *-*-irix6* | *-*-osf* | *-*-hpux11* \
-  | *-*-darwin* | *-*-aix* | *-*-cygwin*)
+  | *-*-darwin* | *-*-aix* | *-*-cygwin* | *-*-mlibc*)
 	# POSIX system.  The OS is supported.
 	config_path="${config_path} posix"
 	;;
diff -urN --no-dereference gcc-clean/libcpp/Makefile.in gcc-workdir/libcpp/Makefile.in
--- gcc-clean/libcpp/Makefile.in
+++ gcc-workdir/libcpp/Makefile.in
@@ -32,10 +32,10 @@
 AUTOHEADER = @AUTOHEADER@
 CATALOGS = $(patsubst %,po/%,@CATALOGS@)
 CC = @CC@
-CFLAGS = @CFLAGS@
+override CFLAGS := @CFLAGS@
 WARN_CFLAGS = @warn@ @c_warn@ @WARN_PEDANTIC@ @WERROR@
 CXX = @CXX@
-CXXFLAGS = @CXXFLAGS@
+override CXXFLAGS := @CXXFLAGS@
 WARN_CXXFLAGS = @warn@ @WARN_PEDANTIC@ @WERROR@
 CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
@@ -45,7 +45,7 @@
 INSTALL_DATA = @INSTALL_DATA@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
-LDFLAGS = @LDFLAGS@
+override LDFLAGS := @LDFLAGS@
 LIBICONV = @LIBICONV@
 LIBINTL = @LIBINTL@
 PACKAGE = @PACKAGE@
diff -urN --no-dereference gcc-clean/libgcc/config.host gcc-workdir/libgcc/config.host
--- gcc-clean/libgcc/config.host
+++ gcc-workdir/libgcc/config.host
@@ -103,7 +103,7 @@
 	;;
 avr-*-*)
 	cpu_type=avr
-	;;    
+	;;
 bfin*-*)
 	cpu_type=bfin
 	;;
@@ -304,6 +304,11 @@
   tmake_file="$tmake_file t-crtstuff-pic t-libgcc-pic t-eh-dw2-dip t-slibgcc t-slibgcc-fuchsia"
   extra_parts="crtbegin.o crtend.o"
   ;;
+*-*-*-mlibc)
+  extra_parts="$extra_parts crtbegin.o crtbeginS.o crtend.o crtendS.o"
+  tmake_file="$tmake_file t-crtstuff-pic"
+  tmake_file="$tmake_file t-slibgcc t-slibgcc-gld t-slibgcc-elf-ver t-libgcc-pic"
+  ;;
 *-*-linux* | frv-*-*linux* | *-*-kfreebsd*-gnu | *-*-gnu* | *-*-kopensolaris*-gnu | *-*-uclinuxfdpiceabi)
   tmake_file="$tmake_file t-crtstuff-pic t-libgcc-pic t-eh-dw2-dip t-slibgcc t-slibgcc-gld t-slibgcc-elf-ver t-linux"
   extra_parts="crtbegin.o crtbeginS.o crtbeginT.o crtend.o crtendS.o"
@@ -439,6 +444,14 @@
 	tmake_file="${tmake_file} ${cpu_type}/t-softfp t-softfp"
 	tmake_file="${tmake_file} t-dfprules"
 	;;
+aarch64*-*-menix*)
+	extra_parts="$extra_parts crtbegin.o crtend.o crti.o crtn.o"
+	extra_parts="$extra_parts crtfastmath.o"
+	md_unwind_header=aarch64/aarch64-unwind.h
+	tmake_file="${tmake_file} ${cpu_type}/t-aarch64"
+	tmake_file="${tmake_file} ${cpu_type}/t-lse t-slibgcc-libgcc"
+	tmake_file="${tmake_file} ${cpu_type}/t-softfp t-softfp t-crtfm"
+	;;
 aarch64*-*-linux*)
 	extra_parts="$extra_parts crtfastmath.o"
 	md_unwind_header=aarch64/linux-unwind.h
@@ -941,6 +954,11 @@
 		extra_parts="$extra_parts vtv_start.o vtv_end.o vtv_start_preinit.o vtv_end_preinit.o"
 	fi
 	;;
+x86_64-*-*-mlibc)
+	extra_parts="$extra_parts crtprec32.o crtprec64.o crtprec80.o crtfastmath.o"
+	tmake_file="$tmake_file i386/t-crtpc t-crtfm i386/t-crtstuff t-dfprules"
+	tmake_file="$tmake_file i386/t-linux"
+	;;
 ia64*-*-elf*)
 	extra_parts="$extra_parts crtbeginS.o crtendS.o crtfastmath.o"
 	tmake_file="ia64/t-ia64 ia64/t-ia64-elf ia64/t-eh-ia64 t-crtfm t-softfp-tf ia64/t-softfp t-softfp ia64/t-softfp-compat"
@@ -998,6 +1016,11 @@
 	tmake_file="${tmake_file} t-crtfm loongarch/t-crtstuff"
 	tmake_file="${tmake_file} t-slibgcc-libgcc"
 	;;
+loongarch*-menix*)
+	extra_parts="$extra_parts crtfastmath.o"
+	tmake_file="${tmake_file} t-crtfm loongarch/t-crtstuff"
+	tmake_file="${tmake_file} t-slibgcc-libgcc"
+	;;
 m32r-*-elf*)
 	tmake_file="$tmake_file m32r/t-m32r t-fdpbit"
 	extra_parts="$extra_parts crtinit.o crtfini.o"
@@ -1347,6 +1370,10 @@
 	tmake_file="${tmake_file} t-softfp-sfdf t-softfp-excl t-softfp t-gnu-prefix pru/t-pru"
 	tm_file="$tm_file pru/pru-abi.h"
 	;;
+riscv*-*-mlibc*)
+    tmake_file="${tmake_file} riscv/t-crtstuff riscv/t-softfp${host_address} t-softfp riscv/t-elf riscv/t-elf${host_address} t-slibgcc-libgcc"
+    extra_parts="$extra_parts crtbegin.o crtend.o crti.o crtn.o crtendS.o crtbeginT.o"
+	;;
 riscv*-*-linux*)
 	tmake_file="${tmake_file} riscv/t-crtstuff riscv/t-softfp${host_address} t-softfp riscv/t-elf riscv/t-elf${host_address} t-slibgcc-libgcc"
 	extra_parts="$extra_parts crtbegin.o crtend.o crti.o crtn.o crtendS.o crtbeginT.o"
diff -urN --no-dereference gcc-clean/libstdc++-v3/configure gcc-workdir/libstdc++-v3/configure
--- gcc-clean/libstdc++-v3/configure
+++ gcc-workdir/libstdc++-v3/configure
@@ -38822,7 +38822,7 @@
 
   fi
     ;;
-  *-linux* | *-uclinux* | *-gnu* | *-kfreebsd*-gnu | *-cygwin* | *-solaris*)
+  *-linux* | *-uclinux* | *-gnu* | *-kfreebsd*-gnu | *-cygwin* | *-solaris* | *-mlibc*)
 
   # All these tests are for C++; save the language and the compiler flags.
   # The CXXFLAGS thing is suspicious, but based on similar bits previously
